src/test/synchronizer.test.ts

```diff
@@ -17,57 +17,165 @@ describe("Synchronizer tests", () => {
     synchronizer = new Synchronizer(issueKit, projectKit, logger);
   });
 
-  test("should fail on invalid event name", async () => {
-    const randomEventName = new Date().toDateString();
-    const expectedError = `Event '${randomEventName}' is not expected. Failing.`;
-    await expect(synchronizer.synchronizeIssue({ eventName: randomEventName, payload: {} })).rejects.toThrow(
-      expectedError,
-    );
-  });
+  describe("synchronize Issues function", () => {
+    test("should fail on invalid event name", async () => {
+      const randomEventName = new Date().toDateString();
+      const expectedError = `Event '${randomEventName}' is not expected. Failing.`;
+      await expect(synchronizer.synchronizeIssue({ eventName: randomEventName, payload: {} })).rejects.toThrow(
+        expectedError,
+      );
+    });
 
-  test("should fail on issue event without payload", async () => {
-    await expect(synchronizer.synchronizeIssue({ eventName: "issues", payload: {} })).rejects.toThrow(
-      "Issue payload object was null",
-    );
-  });
+    test("should fail on issue event without payload", async () => {
+      await expect(synchronizer.synchronizeIssue({ eventName: "issues", payload: {} })).rejects.toThrow(
+        "Issue payload object was null",
+      );
+    });
 
-  test("should log when all issues will be synced", async () => {
-    issueKit.getAllIssues.mockReturnValue(Promise.resolve([]));
-    await synchronizer.synchronizeIssue({ eventName: "workflow_dispatch", payload: {} });
+    test("should log when all issues will be synced", async () => {
+      issueKit.getAllIssues.mockResolvedValue([]);
+      await synchronizer.synchronizeIssue({ eventName: "workflow_dispatch", payload: {} });
 
-    expect(logger.notice).toBeCalledWith("Closed issues will be synced.");
+      expect(logger.notice).toBeCalledWith("Closed issues will be synced.");
+    });
+
+    test("should log when only open issues will be synced", async () => {
+      issueKit.getAllIssues.mockResolvedValue([]);
+      await synchronizer.synchronizeIssue({
+        eventName: "workflow_dispatch",
+        payload: { inputs: { excludeClosed: "true" } },
+      });
+
+      expect(logger.notice).toBeCalledWith("Closed issues will NOT be synced.");
+    });
   });
 
-  test("should log when only open issues will be synced", async () => {
-    issueKit.getAllIssues.mockReturnValue(Promise.resolve([]));
-    await synchronizer.synchronizeIssue({
-      eventName: "workflow_dispatch",
-      payload: { inputs: { excludeClosed: "true" } },
+  describe("update one issue", () => {
+    let nodeData: NodeData;
+    let ctx: GitHubContext;
+    let issueNumber: number;
+    beforeEach(() => {
+      issueNumber = Math.floor(Math.random() * 100);
```

Comment:
Any particular the reason for the random here?

Reply:
I like having to have every test running with an unique id. Adds an extra layer of security over any kind of unseen remaining side effect.

So, if for some reason, the object that uses that value didn't clean up properly, by having a new object we ensure that the previous test won't pass just by having a reference to that old data.

Reply:
I'm more inclined to have reproducible tests, which wouldn't depend on random.  

> So, if for some reason, the object that uses that value didn't clean up properly, by having a new object we ensure that the previous test won't pass just by having a reference to that old data.

Could you elaborate on that? Which object do you mean here, and maybe we can make ourselves sure without having a probability of a flacky test?  

P.S. Also, another rule of thumb that saved my ass a lot: make your tests fail correctly before making them pass. Incorrect cleanup is not the only reason for tests to pass when they shouldn't.

Reply:
> I'm more inclined to have reproducible tests, which wouldn't depend on random.

Fair enough. I changed them now.

> Could you elaborate on that? Which object do you mean here, and maybe we can make ourselves sure without having a probability of a flacky test?

While I am always recreating the object, some integration tests don't do that and have cases where you call a function that changes a state.
So, you have a two functions:
```typescript
const myObject = { abc: "def" };

test1() => {
    saveToDb(myObject);
}

test2() => {
    // for some reason this method failed silently or wasn't executed
   saveToDb(myObject);
   const obj = fetchFromDb();

   expect(obj).toBe(myObject);
}
```

Let's say that `storeToDb` keeps a value in memory and `fetchFromDb` returns the last value saved on memory.
If `test1` works correctly and saves the object to the memory, and, *for some unexpected reason*, when `test2` calls `storeToDb` nothing happens (maybe they wrote an early `return` method), when `test2` calls `fetchFromDb` it will fetch the object from `test1`, which is identical to `test2`, hence making the test succeed, not because it worked, but because all the tests share the same reference.

It's an extremely weird case, but I have seen it happen more times than what I expected and it consumed way too many developer hours so I got used to make sure that all the tests runs with unique objects.

> P.S. Also, another rule of thumb that saved my ass a lot: make your tests fail correctly before making them pass. Incorrect cleanup is not the only reason for tests to pass when they shouldn't.

Omg, that feeling when I break something on purpose and the tests are still passing

![image](https://user-images.githubusercontent.com/8524599/214081764-fa1e21e8-726b-4f20-8df1-6425837d71a0.png)